src/HydraAuction/Platform/Interface.hs

```diff
@@ -0,0 +1,387 @@
+
+instance ToJSON (EntityKind x) where
+  toJSON = Aeson.String . T.pack . show
+
+-- Input/Output
+
+data ClientInput entity
+  = Query (EntityQuery entity)
+  | Command ClientCommand
+
+data ClientCommand
+  = ReportAnnouncedAuction AuctionTerms
+  | ReportBidderApproval BidderApproval
+  | ReportHeadDelegate HydraHeadInfo Actor
+  deriving stock (Generic, Show, Eq, Ord)
+  deriving anyclass (ToJSON, FromJSON)
+
+data ServerOutput entity
+  = CommandResult CommandResult
+  | QueryPerformed (EntityQueryResponse entity)
+
+data CommandResult = EntityCreated | EntityAlreadyExists
+  deriving stock (Generic, Show, Eq, Ord)
+  deriving anyclass (ToJSON, FromJSON)
+
+data EntityQuery entity = MkQuery
+  { filters :: [EntityFilter entity]
+  , limit :: Maybe Int
+  }
+newtype EntityQueryResponse entity = MkResponse [entity]
+
+deriving stock instance Entity entity => Eq (EntityQuery entity)
+deriving stock instance Entity entity => Generic (EntityQuery entity)
+deriving stock instance Entity entity => Show (EntityQuery entity)
+deriving anyclass instance Entity entity => FromJSON (EntityQuery entity)
+deriving anyclass instance Entity entity => ToJSON (EntityQuery entity)
+
+deriving stock instance Entity entity => Eq (ClientInput entity)
+deriving stock instance Entity entity => Generic (ClientInput entity)
+deriving stock instance Entity entity => Show (ClientInput entity)
+deriving anyclass instance Entity entity => FromJSON (ClientInput entity)
+deriving anyclass instance Entity entity => ToJSON (ClientInput entity)
+
+deriving stock instance Entity entity => Eq (ServerOutput entity)
+deriving stock instance Entity entity => Generic (ServerOutput entity)
+deriving stock instance Entity entity => Show (ServerOutput entity)
+deriving anyclass instance Entity entity => FromJSON (ServerOutput entity)
+deriving anyclass instance Entity entity => ToJSON (ServerOutput entity)
+
+deriving stock instance Entity entity => Eq (EntityQueryResponse entity)
+deriving stock instance Entity entity => Generic (EntityQueryResponse entity)
+deriving stock instance Entity entity => Show (EntityQueryResponse entity)
+deriving anyclass instance Entity entity => FromJSON (EntityQueryResponse entity)
+deriving anyclass instance Entity entity => ToJSON (EntityQueryResponse entity)
+
+-- SomeX instances
+
+{-
+EntityKind does not make sense for Command case,
+but this type still requires it anyway.
+This is designed that way because, GADT/existentials break deriving,
+and hiding EntitkyKind existentialy inside EntityQuery will
+require to much more manual instances written.
+-}
+
+data Some (container :: Type -> Type)
+  = forall entity.
+    Entity entity =>
+    MkSome (EntityKind entity) (container entity)
+
+instance
+  (forall entity. Show entity => Show (container entity)) =>
+  Show (Some container)
+  where
+  show (MkSome _ container) =
+    "MkSome (" <> show container <> " )"
+
+parseSomeEntityKindJSON :: Aeson.Value -> Aeson.Parser (Some EntityKind)
```

Comment:
Could this be `Aeson.Value -> Aeson.Parser EntityKind` instead? I think this would also simplify the FromJSON instance where we use this to parse later

Reply:
No, one cannot parse GADT, only print :D
That is like the whole main problem.
Parse function would return not `forall a. EntityKind a`, but `EntityKind x` for some concrete x type.
That is essentially `exists x. EntityKind x`, which in Haskell is emulated with existential datatypes,
`Some ...` in this case.